nomos-libp2p/src/behaviour/nat/address_mapper/protocols/pcp/wire/mod.rs

```diff
@@ -0,0 +1,174 @@
+            Self::Error::InvalidResponse("Address is not IPv4-mapped IPv6 format".to_owned())
+        })
+    }
+}
+
+/// Transport protocol for port mappings (IANA protocol numbers)
+#[derive(Debug, Clone, Copy, PartialEq, Eq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u8)]
+pub enum Protocol {
+    /// TCP protocol (IANA 6)
+    Tcp = 6,
+    /// UDP protocol (IANA 17)
+    Udp = 17,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u8)]
+pub enum ResultCode {
+    Success = 0,
+    UnsupportedVersion = 1,
+    NotAuthorized = 2,
+    MalformedRequest = 3,
+    UnsupportedOpcode = 4,
+    UnsupportedOption = 5,
+    MalformedOption = 6,
+    NetworkFailure = 7,
+    NoResources = 8,
+    UnsupportedProtocol = 9,
+    UserExceededQuota = 10,
+    CannotProvideExternalAddress = 11,
+    AddressMismatch = 12,
+    ExcessiveRemotePeers = 13,
+}
+
+#[derive(Debug, FromBytes, IntoBytes, Unaligned, Immutable, KnownLayout)]
+#[repr(C)]
+pub struct PcpRequestHeader {
+    /// PCP protocol version (must be 2)
+    pub version: u8,
+    /// Operation code (0=ANNOUNCE, 1=MAP, etc.)
+    pub opcode: u8,
+    /// Reserved field (must be 0)
+    pub reserved: [u8; 2],
+    /// Requested lifetime in seconds
+    pub lifetime: U32NE,
+    /// Client IPv4 address as IPv4-mapped IPv6
+    pub client_ip: Ipv4MappedIpv6,
+}
+
+#[derive(Debug, FromBytes, IntoBytes, Unaligned, Immutable)]
+#[repr(C)]
+pub struct PcpRequest<Payload> {
+    pub header: PcpRequestHeader,
+    pub payload: Payload,
+}
+
+#[derive(Debug, FromBytes, IntoBytes, Unaligned, Immutable, KnownLayout)]
+#[repr(C)]
+pub struct PcpResponseHeader {
+    /// PCP protocol version (always 2)
+    pub version: u8,
+    /// Operation code with response bit (0x80 | opcode)
+    pub opcode: u8,
+    /// Result code (0=success, see `ResultCode` enum)
+    pub result_code: u8,
+    /// Reserved field
+    pub reserved: u8,
+    /// Server-assigned lifetime in seconds
+    pub lifetime: U32NE,
+    /// Server epoch time in seconds since boot
+    pub epoch_time: U32NE,
+    /// Reserved field (12 bytes)
+    pub reserved2: [u8; 12],
+}
+
+impl PcpResponseHeader {
+    /// Extracts the base opcode from the response opcode field.
+    ///
+    /// Response opcodes have the high bit (0x80) set, this method
+    /// masks it out to get the original request opcode.
+    pub const fn base_opcode(&self) -> u8 {
+        self.opcode & 0x7F
+    }
+
+    /// Creates a response opcode from a request opcode by setting the response
+    /// bit.
+    pub const fn make_response_opcode(request_opcode: u8) -> u8 {
+        request_opcode | 0x80
+    }
+}
+
+#[derive(Debug, FromBytes, IntoBytes, Unaligned, Immutable, KnownLayout)]
+#[repr(C)]
+pub struct PcpResponse<Payload> {
+    pub header: PcpResponseHeader,
+    pub payload: Payload,
+}
+
+impl<Payload> PcpRequest<Payload> {
+    pub fn new_with_header(client_addr: Ipv4Addr, opcode: u8, lifetime: u32) -> Self
```

Comment:
Question: is this somehow required by the `zerocopy` crate? Why can't we have a function that takes the payload and returns a fully composed request instead of a request with a zeroed payload that needs filling afterwards?